src/server/input/key_repeat_dispatcher.cpp

```diff
@@ -158,16 +158,19 @@ bool mi::KeyRepeatDispatcher::dispatch(std::shared_ptr<MirEvent const> const& ev
 // Returns true if the original event has been handled, that is ::dispatch should not pass it on.
 bool mi::KeyRepeatDispatcher::handle_key_input(MirInputDeviceId id, MirKeyboardEvent const* kev)
 {
-    std::lock_guard<std::mutex> lg(repeat_state_mutex);
-    auto& device_state = ensure_state_for_device_locked(lg, id);
-
     auto scan_code = mir_keyboard_event_scan_code(kev);
 
     switch (mir_keyboard_event_action(kev))
     {
     case mir_keyboard_action_up:
     {
-        device_state.repeat_alarm.reset();
+        std::shared_ptr<time::Alarm> alarm;
+        {
+            std::lock_guard<std::mutex> lg(repeat_state_mutex);
+            auto& device_state = ensure_state_for_device_locked(lg, id);
+            alarm = std::move(device_state.repeat_alarm);
+        }
+        alarm.reset();
```

Comment:
Looks like there's a repeating pattern here. And one that could be further generalized to:
```c++
auto const exchange_alarm_for_device = [this](MirInputDeviceId id, std::shared_ptr<time::Alarm>& alarm)
    {
            std::lock_guard<std::mutex> lg(repeat_state_mutex);
            std::swap(alarm, ensure_state_for_device_locked(lg, id).repeat_alarm);
    };
```
Which could also cover the setting case.

Reply:
I came to the conclusion `ensure_state_for_device_locked()` is the wrong abstraction. Since we only use it in the context of setting the repeat alarm, I've changed it to `set_repeat_alarm()`.